rfcs/20210115-tfx-periodic-training.md

```diff
@@ -0,0 +1,445 @@
+pipeline:*
+
+<div style="text-align:center"><img src='20210115-tfx-periodic-training/new_version.png', width='500'></div>
+
+Note that the normal run and backfill run need to have the same pipeline name
+for correct ML metadata tracking.
+
+##### Partial pipeline execution
+
+In certain backfill cases users may not want to run the entire pipeline. For
+example, when there are multiple spans in the rolling window that need to be
+backfilled, it’s not necessary to retrain the model multiple times. In such
+cases, user might want to (conceptually) divide the pipeline in two for a faster
+backfill:
+
+*   The first portion contains components like ExampleGen, StatisticsGen and
+    ExampleValidator. This part of the pipeline is scheduled for each Span.
+*   The second portion contains a Resolver plus components like Transform,
+    Trainer, Evaluator and Pusher. This part of the pipeline only needs to run
+    once for updating the model after all Spans are updated.
+
+We plan to provide the ability to specify a partial pipeline to execute at
+runtime (see Future Work below) to enable this workflow.
+
+An alternative approach is to define the two pipeline portions as TFX pipelines
+in their own right, to make backfill easier. This is not recommended because it
+makes the normal case of periodic training more brittle: the two pipelines must
+now be triggered in sequence by some non-TFX mechanism, thereby multiplying the
+possible fault conditions.
+
+##### Training Considerations for Backfill Runs
+
+When there are issues with input data which require new versions of a span, the
+models that have been trained with the problematic data may be of low quality.
+Warmstarting can propagate this damage forward in an unpredictable way. To
+recover a pipeline that uses warmstarting, it’s advisable to retrain from
+scratch. If full retraining is too costly, consider manually picking up a valid
+previous model to warmstart using
+[Importer](https://github.com/tensorflow/tfx/blob/r0.26.0/tfx/components/common_nodes/importer_node.py).
+
+#### ExampleGen Data Selection
+
+As mentioned above, each ExampleGen execution generates only one Span at a time.
+ExampleGen can be configured to process a specific Span (always taking its
+latest Version); if a Span is not specified, ExampleGen selects the input data
+corresponding to the latest Version of the latest Span by default. The latest
+here means the one that has the highest Span id.
+
+Relying on “latest Span” semantics in periodic automation introduces a new
+failure mode: if the data isn’t available as expected the pipeline will
+reprocess old data. While it’s possible to guard against this with upstream
+monitoring and alerting, we recommend specifying the span as runtime parameter
+when triggering the run. For example, for a daily cron-triggered pipeline, pass
+the scheduled date to the job for calculating target span of ExampleGen. The
```

Comment:
Is that even possible? Currently RuntimeParameters are only supported in Kubeflow and I believe Kubeflow does not allow for dynamic parameters when using [run triggers](https://www.kubeflow.org/docs/pipelines/overview/concepts/run-trigger/). How do you set the scheduled date in Kubeflow? Is there plans to add support for it on Airflow?

Reply:
For beam based DagRunner, it's main job is "compile and run" so it's by default runtime param.
For Kubeflow, the runtime params can be specified in UI or using Client API. I will confirm if run triggers supports it.
For Airflow, it has its own [runtime parameters](https://stackoverflow.com/questions/53663534/for-apache-airflow-how-can-i-pass-the-parameters-when-manually-trigger-dag-via) supports,

Reply:
> For Airflow, it has its own runtime parameters supports
Indeed, wondering if we can add support for those from TFX (use RuntimeParameter similarly). This could be accomplished by using template_fields in the operator and mapping RuntimParamete to be [DagParam]( https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/dagparam/index.html#airflow.models.dagparam.DagParam) object (only Airflow 2.0>)
> Kubeflow
run triggers does support parametrization, but it does not support having a parameter be the current date/execution date as far as I know.

Reply:
> > For Airflow, it has its own runtime parameters supports
> 
> Indeed, wondering if we can add support for those from TFX (use RuntimeParameter similarly). This could be accomplished by using template_fields in the operator and mapping RuntimParamete to be [DagParam](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/dagparam/index.html#airflow.models.dagparam.DagParam) object (only Airflow 2.0>)

Not Airflow expert, but I wonder if we can directly use the params in airflow pipeline dsl?

e.g. ExampleGen(span={some airflow runtime param})

> 
> > Kubeflow
> 
> run triggers does support parametrization, but it does not support having a parameter be the current date/execution date as far as I know.

Reply:
> Not Airflow expert, but I wonder if we can directly use the params in airflow pipeline dsl?

They won't. Airflow be default resolves `template_fields` which are class variables `self.X`. Current implementation just wraps everything into a [function](https://github.com/tensorflow/tfx/blob/master/tfx/orchestration/airflow/airflow_component.py#L118) which means that Airflow has no visibility into it. 

One way to do it would be to add all exec_properties as a class variable, mark it as template_field and set all exec_properties there in a list. It may work, would need to test this.

Reply:
> > For Airflow, it has its own runtime parameters supports
> 
> Indeed, wondering if we can add support for those from TFX (use RuntimeParameter similarly). This could be accomplished by using template_fields in the operator and mapping RuntimParamete to be [DagParam](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/models/dagparam/index.html#airflow.models.dagparam.DagParam) object (only Airflow 2.0>)
> 
> > Kubeflow
> 
> run triggers does support parametrization, but it does not support having a parameter be the current date/execution date as far as I know.

I see, Note that in Managed Pipeline we will for sure have a placeholder for the date/time.
For kubeflow run trigger, if it's not supported, a workaround can be we use client API to pass in date/span in cli, and you can have a script/cron that triggers CLI periodically.

Reply:
> I see, Note that in Managed Pipeline we will for sure have a placeholder for the date/time.
For kubeflow run trigger, if it's not supported, a workaround can be we use client API to pass in date/span in cli, and you can have a script/cron that triggers CLI periodically.

That's true. Still not ideal though. Wondering if anyone from Kubeflow can weigh in here.

Reply:
> > I see, Note that in Managed Pipeline we will for sure have a placeholder for the date/time.
> > For kubeflow run trigger, if it's not supported, a workaround can be we use client API to pass in date/span in cli, and you can have a script/cron that triggers CLI periodically.
> 
> That's true. Still not ideal though. Wondering if anyone from Kubeflow can weigh in here.

Hi, I asked Kubeflow team, they do have that,  here is an [example](https://github.com/kubeflow/pipelines/blob/34d23aa924720ead13fe67ebca5c1ab9926585ee/backend/src/crd/samples/scheduledworkflow/parameterized.yaml#L29-L34)

Reply:
Thats awesome! Thanks! Would be great if we could include this reference on the RFC or public docs as an example on how to do scheduled training.

Reply:
> default='[[ScheduledTime]]'

![Screen Shot 2021-01-29 at 3 20 03 PM](https://user-images.githubusercontent.com/46733165/106337235-86dfa000-6245-11eb-8e49-f149c644df20.png)
you can also change "hello world" to placeholder like "[[ScheduledTime]]" in UI, if it's not specified in UI, it will use default value like the code you wrote above